src/micro.c

```diff
@@ -0,0 +1,680 @@
+        micro_update_last_error(ep, err);
+        microError_Destroy(err);
+    }
+
+    // <>/<> TODO: Should we stop the service? The Go client does.
+    microService_Stop(m);
+
+    if (m->prev_on_error != NULL)
+    {
+        (*m->prev_on_error)(nc, sub, s, m->prev_on_error_closure);
+    }
+}
+
+static microError *
+wrap_connection_event_callbacks(microService *m)
+{
+    natsStatus s = NATS_OK;
+
+    if (m == NULL)
+        return micro_ErrorInvalidArg;
+
+    IFOK(s, natsConn_getClosedCallback(&m->prev_on_connection_closed, &m->prev_on_connection_closed_closure, m->nc));
+    IFOK(s, natsConn_setClosedCallback(m->nc, on_connection_closed, m));
+
+    IFOK(s, natsConn_getDisconnectedCallback(&m->prev_on_connection_disconnected, &m->prev_on_connection_disconnected_closure, m->nc));
+    IFOK(s, natsConn_setDisconnectedCallback(m->nc, on_connection_disconnected, m));
+
+    IFOK(s, natsConn_getErrorCallback(&m->prev_on_error, &m->prev_on_error_closure, m->nc));
+    IFOK(s, natsConn_setErrorCallback(m->nc, on_error, m));
+
+    return microError_Wrapf(micro_ErrorFromStatus(s), "failed to wrap connection event callbacks");
+}
+
+static microError *
+unwrap_connection_event_callbacks(microService *m)
+{
+    natsStatus s = NATS_OK;
+
+    if (m == NULL)
+        return micro_ErrorInvalidArg;
+
+    IFOK(s, natsConn_setClosedCallback(m->nc, m->prev_on_connection_closed, m->prev_on_connection_closed_closure));
+    IFOK(s, natsConn_setDisconnectedCallback(m->nc, m->prev_on_connection_disconnected, m->prev_on_connection_disconnected_closure));
+    IFOK(s, natsConn_setErrorCallback(m->nc, m->prev_on_error, m->prev_on_error_closure));
+
+    return microError_Wrapf(micro_ErrorFromStatus(s), "failed to unwrap connection event callbacks");
+}
+
+microError *
+microService_AddGroup(microGroup **new_group, microService *m, const char *prefix)
+{
+    if ((m == NULL) || (new_group == NULL) || (prefix == NULL))
+        return micro_ErrorInvalidArg;
+
+    *new_group = NATS_CALLOC(1, sizeof(microGroup) +
+                                    strlen(prefix) + 1); // "prefix.""
+    if (new_group == NULL)
+    {
+        return micro_ErrorOutOfMemory;
+    }
+
+    memcpy((*new_group)->prefix, prefix, strlen(prefix) + 1);
+    (*new_group)->m = m;
+    (*new_group)->next = m->groups;
+    m->groups = *new_group;
+
+    return NULL;
+}
+
+microError *
+microGroup_AddGroup(microGroup **new_group, microGroup *parent, const char *prefix)
+{
+    char *p;
+    int len;
+
+    if ((parent == NULL) || (new_group == NULL) || (prefix == NULL))
+        return micro_ErrorInvalidArg;
+
+    *new_group = NATS_CALLOC(1, sizeof(microGroup) +
+                                    strlen(parent->prefix) + 1 + // "parent_prefix."
+                                    strlen(prefix) + 1);         // "prefix\0"
+    if (new_group == NULL)
+    {
+        return micro_ErrorOutOfMemory;
+    }
+
+    p = (*new_group)->prefix;
+    len = strlen(parent->prefix);
```

Comment:
On windows you would get:
```
warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
```
So cast to `(int)`.

Reply:
I changed all variables used for (str)lengths to `size_t`, it more appropriate anyway. `#backfromgoland`.